File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.py
def solution(cap, n, deliveries, pickups):
    answer = 0
    distance=n
    i, j=n-1, n-1
    
    while i>=0 or j>=0:
        deliver_cap=cap
        pickup_cap=cap
        
        if deliveries[i]!=0 or pickups[j]!=0:
            answer+=distance*2
            while deliver_cap>0:
                while deliveries[i]==0 and i!=-1:
                    i-=1
                deliveries[i]-=1
                deliver_cap-=1
                
            while pickup_cap>0:
                while pickups[j]==0 and j!=-1:
                    j-=1
                pickups[j]-=1
                pickup_cap-=1
                
            while deliveries[i]==0 and i!=-1:
                i-=1
            while pickups[j]==0 and j!=-1:
                j-=1
                
        else:
            i-=1
            j-=1
        distance=max(i,j)+1
    
    return answer

File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.java

import java.util.*;

class Solution {

    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        long answer =0;
        Stack<Integer> deliver = new Stack<>(); //배달 상자
        Stack<Integer> pick = new Stack<>(); // 수거 상자

        for (int i=0; i<deliveries.length; i++){
            if(deliveries[i]>0) {
                deliver.add(i);
            }

            if(pickups[i]>0){
                pick.add(i);
            }
        }

        while(!deliver.isEmpty() || !pick.isEmpty()){
            //각자 cap씩 할당
            int dcap=cap;
            int pcap=cap;
            int max=0;


            int deliverAddress =0;
            while(dcap>0 && !deliver.isEmpty()){
                int idx=deliver.pop();
                deliverAddress = Math.max(deliverAddress,idx);

                if(dcap<deliveries[idx]){
                    deliveries[idx]-=dcap;
                    deliver.push(idx);
                    dcap=0;
                } else {
                    dcap-=deliveries[idx];
                    deliveries[idx]=0;
                }
            }

            int pickAddress=0;
            while(pcap>0 && !pick.isEmpty()){
                int idx=pick.pop();
                pickAddress = Math.max(pickAddress,idx);

                if(pcap<pickups[idx]){
                    pickups[idx]-=pcap;
                    pick.push(idx);
                    pcap=0;
                } else {
                    pcap-=pickups[idx];
                    pickups[idx]=0;
                }
            }

            max = Math.max(deliverAddress,pickAddress);
            answer+=(max+1)*2;

        }

        System.out.println(answer);
        return answer;

    }
    public static void main(String[] args) {
        Solution main = new Solution();
        main.solution(2, 7, new int[]{1, 0, 2, 0, 1, 0, 2}, new int[]{0, 2, 0, 1, 0, 2, 0});
    }
}




